tedana/utils/utils.py

```diff
@@ -1,198 +1,452 @@
+        Whether to copy header from `ref_img` to new image. Default: True
+    copy_meta : bool, optional
+        Whether to copy meta from `ref_img` to new image. Default: False
 
     Returns
     -------
-    ndarray
-        Masked array of shape (nx*ny*nz[, Ne[, nt]])
+    gii : nib.gifti.GiftiImage
+        GiftiImage
     """
 
-    s = data.shape
+    if isinstance(ref_img, str):
+        ref_img = nib.load(ref_img)
 
-    N = s[0] * s[1] * s[2]
-    news = []
-    news.append(N)
+    if data.ndim == 1:
+        data = np.atleast_2d(data).T
 
-    if len(s) > 3:
-        news.extend(s[3:])
+    darrays = [make_gii_darray(ref_img.darrays[n], d, copy_meta=copy_meta)
+               for n, d in enumerate(data.T)]
+    gii = nib.gifti.GiftiImage(header=ref_img.header if copy_header else None,
+                               extra=ref_img.extra,
+                               meta=ref_img.meta if copy_meta else None,
+                               labeltable=ref_img.labeltable,
+                               darrays=darrays)
 
-    tmp1 = np.reshape(data, news)
-    fdata = tmp1.compress((mask > 0).ravel(), axis=0)
+    return gii
 
-    return fdata.squeeze()
 
+def make_gii_darray(ref_array, data, copy_meta=False):
+    """
+    Converts `data` into GiftiDataArray format like `ref_array`
 
-def unmask(data, mask):
+    Parameters
+    ----------
+    ref_array : str or img_like
+        Reference array
+    data : (S,) array_like
+        Data to be saved
+    copy_meta : bool, optional
+        Whether to copy meta from `ref_img` to new image. Default: False
+
+    Returns
+    -------
+    gii : nib.gifti.GiftiDataArray
+        Output data array instance
     """
-    Unmasks `data` using non-zero entries of `mask`
+
+    if not isinstance(ref_array, nib.gifti.GiftiDataArray):
+        raise TypeError('Provided reference is not a GiftiDataArray.')
+    darray = nib.gifti.GiftiDataArray(data,
+                                      intent=ref_array.intent,
+                                      datatype=data.dtype,
+                                      encoding=ref_array.encoding,
+                                      endian=ref_array.endian,
+                                      coordsys=ref_array.coordsys,
+                                      ordering=ref_array.ind_ord,
+                                      meta=ref_array.meta if copy_meta else None)
+
+    return darray
+
+
+def uncat2echos(data):
+    """
+    Combines Z- and echo-axis in `data`
 
     Parameters
     ----------
-    data : array_like
-        Masked array of shape (nx*ny*nz[, Ne[, nt]])
-    mask : array_like
-        Boolean array of shape (nx, ny, nz)
+    data : (X x Y x Z x E x T) array_like
+        Multi-echo data array
 
     Returns
     -------
-    ndarray
-        Array of shape (nx, ny, nz[, Ne[, nt]])
+    fdata : (X x Y x M x T) np.ndarray
+        Z-concatenated multi-echo data array, where M is Z * number of echos
```

Comment:
If we drop zcat compatibility (as discussed in #26), will we drop this function as well ? For now, it's probably worth keeping, but thinking ahead....

Reply:
The `uncat2echos` function is actually no longer needed, I think -- even with retained zcat compatibility! A combination of `load_data()` and `filewrite()` should be enough to ensure we can handle everything. I'll delete this code to remove any confusion!